1.4.2版本，分支注册时候，报maximum number of expressions in a list is 1000 ，和 #2775 不一样

- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
1.4.2版本，分支注册时候，报maximum number of expressions in a list is 1000 ，这个和 #2775 不一样的，这里报错是检查锁的时候报错
io.seata.core.exception.TransactionException: branch register request failed. xid=10.200.203.31:38098:8620051265334516279, msg=ORA-01795: maximum number of expressions in a list is 1000 
 
	at io.seata.server.AbstractTCInboundHandler$4.execute(AbstractTCInboundHandler.java:187) ~[classes/:na] 
	at io.seata.server.AbstractTCInboundHandler$4.execute(AbstractTCInboundHandler.java:179) ~[classes/:na] 
	at io.seata.core.exception.AbstractExceptionHandler.exceptionHandleTemplate(AbstractExceptionHandler.java:124) ~[seata-core-1.4.2.jar:na] 
	at io.seata.server.AbstractTCInboundHandler.handle(AbstractTCInboundHandler.java:179) [classes/:na] 
	at io.seata.core.protocol.transaction.BranchRegisterRequest.handle(BranchRegisterRequest.java:136) [seata-core-1.4.2.jar:na] 
	at io.seata.server.coordinator.DefaultCoordinator.onRequest(DefaultCoordinator.java:473) [classes/:na] 
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.onRequestMessage(ServerOnRequestProcessor.java:124) [seata-core-1.4.2.jar:na] 
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.process(ServerOnRequestProcessor.java:77) [seata-core-1.4.2.jar:na] 
	at io.seata.core.rpc.netty.AbstractNettyRemoting.lambda$processMessage$2(AbstractNettyRemoting.java:279) [seata-core-1.4.2.jar:na] 
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_212] 
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_212] 
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-all-4.1.30.Final.jar:4.1.30.Final] 
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_212] 
Caused by: io.seata.common.exception.StoreException: ORA-01795: maximum number of expressions in a list is 1000 
 
	at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.acquireLock(LockStoreDataBaseDAO.java:182) ~[classes/:na] 
	at io.seata.server.storage.db.lock.DataBaseLocker.acquireLock(DataBaseLocker.java:59) ~[classes/:na] 
	at io.seata.server.lock.AbstractLockManager.acquireLock(AbstractLockManager.java:59) ~[classes/:na] 
	at io.seata.server.session.BranchSession.lock(BranchSession.java:278) ~[classes/:na] 
	at io.seata.server.transaction.at.ATCore.branchSessionLock(ATCore.java:46) ~[classes/:na] 
	at io.seata.server.coordinator.AbstractCore.lambda$branchRegister$0(AbstractCore.java:83) ~[classes/:na] 
	at io.seata.server.storage.db.session.DataBaseSessionManager.lockAndExecute(DataBaseSessionManager.java:194) ~[classes/:na] 
	at io.seata.server.session.SessionHolder.lockAndExecute(SessionHolder.java:331) ~[classes/:na] 
	at io.seata.server.coordinator.AbstractCore.branchRegister(AbstractCore.java:77) ~[classes/:na] 
	at io.seata.server.coordinator.DefaultCore.branchRegister(DefaultCore.java:100) ~[classes/:na] 
	at io.seata.server.coordinator.DefaultCoordinator.doBranchRegister(DefaultCoordinator.java:199) [classes/:na] 
	at io.seata.server.AbstractTCInboundHandler$4.execute(AbstractTCInboundHandler.java:184) ~[classes/:na] 
	... 12 common frames omitted 
Caused by: java.sql.SQLSyntaxErrorException: ORA-01795: maximum number of expressions in a list is 1000 
 
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:509) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:461) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1104) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:550) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:268) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:655) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:270) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:91) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:807) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:983) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1168) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3666) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.T4CPreparedStatement.executeInternal(T4CPreparedStatement.java:1426) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:3713) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1167) ~[ojdbc8-19.3.0.0.jar:19.3.0.0.0] 
	at com.alibaba.druid.pool.DruidPooledPreparedStatement.executeQuery(DruidPooledPreparedStatement.java:227) ~[druid-1.1.23.jar:1.1.23] 
	at io.seata.server.storage.db.lock.LockStoreDataBaseDAO.acquireLock(LockStoreDataBaseDAO.java:126) ~[classes/:na] 
	... 23 common frames omitted 
看报错位置是这里
![1628819578368_0588A5FB-5B7F-4c88-9007-488F419432EF](https://user-images.githubusercontent.com/48412175/129299067-2f2558c5-900a-4974-a7ba-193d9bcc7ada.png)

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```


### Ⅲ. Describe what you expected to happen


### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1.代码结构类似这样
@GlobalTransactional
public void A(){
	service.B()
}

@Transactional
public void B(){
	//删除1000多条
	//添加1000多条
}
2.主要是在一个事务提交中，操作一千多条数据，checklock的时候，in 后面会有1000多个数据，会报错

### Ⅴ. Anything else we need to know?


### Ⅵ. Environment:

- JDK version : 1.8
- OS : centos7
- Others: seata1.4.2

--- a/changes/1.5.0.md
+++ b/changes/1.5.0.md
@@ -126,6 +126,7 @@ Seata 是一款开源的分布式事务解决方案，提供高性能和简单�
   - [[#3912](https://github.com/seata/seata/pull/3912)] 支持通过env配置JVM参数
   - [[#3939](https://github.com/seata/seata/pull/3939)] 使用map优化大量的判断代码
   - [[#3955](https://github.com/seata/seata/pull/3955)] 添加启动banner
+  - [[#3946](https://github.com/seata/seata/pull/3946)] 修改由于修改记录过多导致分支注册及lock释放失败的问题 
   - [[#3949](https://github.com/seata/seata/pull/3949)] `nacos-config.py` 支持默认参数和选择性输入参数
   - [[#3954](https://github.com/seata/seata/pull/3954)] 移除对druid依赖中过期方法的调用
   - [[#3981](https://github.com/seata/seata/pull/3981)] 优化服务端口的优先级设置

--- a/changes/en-us/1.5.0.md
+++ b/changes/en-us/1.5.0.md
@@ -59,6 +59,7 @@
   - [[#3901](https://github.com/seata/seata/pull/3901)] fix seataio/seata-server servlet-api conflict
   - [[#3931](https://github.com/seata/seata/pull/3931)] fix the wrong path and filename when dump the jvm memory for analysis
   - [[#3976](https://github.com/seata/seata/pull/3976)] fix NPE cause by future timeout
+  - [[#3946](https://github.com/seata/seata/pull/3946)] fix register branch and release lock failed when the size of rows that modified is greater than 1000 in oracle 
   - [[#3949](https://github.com/seata/seata/pull/3949)] fix the problem that `nacos-config.py` will not skip blank options. fix bug that split options may cause content loss
   - [[#3988](https://github.com/seata/seata/pull/3988)] fix the problem that nacos not found user when password has special characters
   - [[#3998](https://github.com/seata/seata/pull/3998)] fix the NPE of jedis multi.exec

--- a/core/src/main/java/io/seata/core/store/db/sql/lock/AbstractLockStoreSql.java
+++ b/core/src/main/java/io/seata/core/store/db/sql/lock/AbstractLockStoreSql.java
@@ -15,6 +15,9 @@
  */
 package io.seata.core.store.db.sql.lock;
 
+import java.util.ArrayList;
+import java.util.List;
+
 import io.seata.common.exception.NotSupportYetException;
 import io.seata.config.Configuration;
 import io.seata.config.ConfigurationFactory;
@@ -29,6 +32,8 @@ import io.seata.core.constants.ServerTableColumnsName;
  * @since 1.2.0
  */
 public class AbstractLockStoreSql implements LockStoreSql {
+
+    private static final int MAX_IN_SIZE = 1000;
 
     /**
      * The constant CONFIG.
@@ -50,6 +55,17 @@ public class AbstractLockStoreSql implements LockStoreSql {
     protected static final String IN_PARAMS_PLACE_HOLD = " #in_params# ";
 
     /**
+     * The constant LOCK_TABLE_PK_WHERE_CONDITION_PLACE_HOLD.
+     */
+    protected static final String LOCK_TABLE_PK_WHERE_CONDITION_PLACE_HOLD = " #lock_table_pk_where_condition# ";
+
+    /**
+     * The constant LOCK_TABLE_BRANCH_ID_WHERE_CONDITION_PLACE_HOLD.
+     */
+    protected static final String LOCK_TABLE_BRANCH_ID_WHERE_CONDITION_PLACE_HOLD = " #lock_table_branch_id_where_condition# ";
+
+    
+    /**
      * The constant ALL_COLUMNS.
      * xid, transaction_id, branch_id, resource_id, table_name, pk, row_key, gmt_create, gmt_modified
      */
@@ -70,7 +86,7 @@ public class AbstractLockStoreSql implements LockStoreSql {
      * The constant BATCH_DELETE_LOCK_SQL.
      */
     private static final String BATCH_DELETE_LOCK_SQL = "delete from " + LOCK_TABLE_PLACE_HOLD
-        + " where " + ServerTableColumnsName.LOCK_TABLE_XID + " = ? and " + ServerTableColumnsName.LOCK_TABLE_ROW_KEY + " in (" + IN_PARAMS_PLACE_HOLD + ") ";
+        + " where " + ServerTableColumnsName.LOCK_TABLE_XID + " = ? and (" + LOCK_TABLE_PK_WHERE_CONDITION_PLACE_HOLD + ") ";
 
     /**
      * The constant BATCH_DELETE_LOCK_BY_BRANCH_SQL.
@@ -88,7 +104,7 @@ public class AbstractLockStoreSql implements LockStoreSql {
      * The constant BATCH_DELETE_LOCK_BY_BRANCHS_SQL.
      */
     private static final String BATCH_DELETE_LOCK_BY_BRANCHS_SQL = "delete from " + LOCK_TABLE_PLACE_HOLD
-        + " where " + ServerTableColumnsName.LOCK_TABLE_XID + " = ? and " + ServerTableColumnsName.LOCK_TABLE_BRANCH_ID + " in (" + IN_PARAMS_PLACE_HOLD + ") ";
+        + " where " + ServerTableColumnsName.LOCK_TABLE_XID + " = ? and (" + LOCK_TABLE_BRANCH_ID_WHERE_CONDITION_PLACE_HOLD + ") ";
 
 
     /**
@@ -101,7 +117,7 @@ public class AbstractLockStoreSql implements LockStoreSql {
      * The constant CHECK_LOCK_SQL.
      */
     private static final String CHECK_LOCK_SQL = "select " + ALL_COLUMNS + " from " + LOCK_TABLE_PLACE_HOLD
-        + " where " + ServerTableColumnsName.LOCK_TABLE_ROW_KEY + " in (" + IN_PARAMS_PLACE_HOLD + ")"
+        + " where " + LOCK_TABLE_PK_WHERE_CONDITION_PLACE_HOLD
         + " order by status desc ";
 
     /**
@@ -127,9 +143,11 @@ public class AbstractLockStoreSql implements LockStoreSql {
     }
 
     @Override
-    public String getBatchDeleteLockSql(String lockTable, String paramPlaceHold) {
-        return BATCH_DELETE_LOCK_SQL.replace(LOCK_TABLE_PLACE_HOLD, lockTable).replace(IN_PARAMS_PLACE_HOLD,
-            paramPlaceHold);
+    public String getBatchDeleteLockSql(String lockTable, int rowSize) {
+        List<String> pkNameList = new ArrayList<>();
+        pkNameList.add(ServerTableColumnsName.LOCK_TABLE_ROW_KEY);
+        String whereCondition = buildWhereConditionByPKs(pkNameList,rowSize,MAX_IN_SIZE);
+        return BATCH_DELETE_LOCK_SQL.replace(LOCK_TABLE_PLACE_HOLD, lockTable).replace(LOCK_TABLE_PK_WHERE_CONDITION_PLACE_HOLD, whereCondition);
     }
 
     @Override
@@ -138,9 +156,12 @@ public class AbstractLockStoreSql implements LockStoreSql {
     }
 
     @Override
-    public String getBatchDeleteLockSqlByBranchs(String lockTable, String paramPlaceHold) {
-        return BATCH_DELETE_LOCK_BY_BRANCHS_SQL.replace(LOCK_TABLE_PLACE_HOLD, lockTable).replace(IN_PARAMS_PLACE_HOLD,
-            paramPlaceHold);
+    public String getBatchDeleteLockSqlByBranchs(String lockTable, int branchSize) {
+        List<String> pkNameList = new ArrayList<>();
+        pkNameList.add(ServerTableColumnsName.BRANCH_TABLE_BRANCH_ID);
+        String whereCondition = buildWhereConditionByPKs(pkNameList,branchSize,MAX_IN_SIZE);
+        return BATCH_DELETE_LOCK_BY_BRANCHS_SQL.replace(LOCK_TABLE_PLACE_HOLD, lockTable).replace(LOCK_TABLE_BRANCH_ID_WHERE_CONDITION_PLACE_HOLD,
+            whereCondition);
     }
 
     @Override
@@ -149,13 +170,75 @@ public class AbstractLockStoreSql implements LockStoreSql {
     }
 
     @Override
-    public String getCheckLockableSql(String lockTable, String paramPlaceHold) {
-        return CHECK_LOCK_SQL.replace(LOCK_TABLE_PLACE_HOLD, lockTable).replace(IN_PARAMS_PLACE_HOLD, paramPlaceHold);
+    public String getCheckLockableSql(String lockTable, int rowSize) {
+        List<String> pkNameList = new ArrayList<>();
+        pkNameList.add(ServerTableColumnsName.LOCK_TABLE_ROW_KEY);
+        String whereCondition = buildWhereConditionByPKs(pkNameList,rowSize,MAX_IN_SIZE);
+        return CHECK_LOCK_SQL.replace(LOCK_TABLE_PLACE_HOLD, lockTable).replace(LOCK_TABLE_PK_WHERE_CONDITION_PLACE_HOLD, whereCondition);
     }
 
     @Override
     public String getBatchUpdateStatusLockByGlobalSql(String lockTable) {
         return BATCH_UPDATE_STATUS_LOCK_BY_GLOBAL_SQL.replace(LOCK_TABLE_PLACE_HOLD, lockTable);
     }
+
+    /**
+     * each pk is a condition.the result will like :" (id,userCode) in ((?,?),(?,?)) or (id,userCode) in ((?,?),(?,?)
+     * ) or (id,userCode) in ((?,?))"
+     * Build where condition by pks string.
+     *
+     * @param pkNameList pk column name list
+     * @param rowSize    the row size of records
+     * @param maxInSize  the max in size
+     * @return return where condition sql string.the sql can search all related records not just one.
+     */
+    private String buildWhereConditionByPKs(List<String> pkNameList, int rowSize, int maxInSize) {
+        StringBuilder whereStr = new StringBuilder();
+        //we must consider the situation of composite primary key
+        int batchSize = rowSize % maxInSize == 0 ? rowSize / maxInSize : (rowSize / maxInSize) + 1;
+        for (int batch = 0; batch < batchSize; batch++) {
+            if (batch > 0) {
+                whereStr.append(" or ");
+            }
+            if (pkNameList.size() > 1) {
+                whereStr.append("(");
+            }
+            for (int i = 0; i < pkNameList.size(); i++) {
+                if (i > 0) {
+                    whereStr.append(",");
+                }
+                whereStr.append(pkNameList.get(i));
+            }
+            if (pkNameList.size() > 1) {
+                whereStr.append(")");
+            }
+            whereStr.append(" in ( ");
+
+            int eachSize = (batch == batchSize - 1) ? (rowSize % maxInSize == 0 ? maxInSize : rowSize % maxInSize)
+                : maxInSize;
+            for (int i = 0; i < eachSize; i++) {
+                //each row is a bracket
+                if (i > 0) {
+                    whereStr.append(",");
+                }
+                if (pkNameList.size() > 1) {
+                    whereStr.append("(");
+                }
+                for (int x = 0; x < pkNameList.size(); x++) {
+                    if (x > 0) {
+                        whereStr.append(",");
+                    }
+                    whereStr.append("?");
+                }
+                if (pkNameList.size() > 1) {
+                    whereStr.append(")");
+                }
+            }
+            whereStr.append(" )");
+        }
+
+        return whereStr.toString();
+    }
+
 
 }

--- a/core/src/main/java/io/seata/core/store/db/sql/lock/LockStoreSql.java
+++ b/core/src/main/java/io/seata/core/store/db/sql/lock/LockStoreSql.java
@@ -52,10 +52,10 @@ public interface LockStoreSql {
      * Get batch delete lock sql string.
      *
      * @param lockTable      the lock table
-     * @param paramPlaceHold the param place hold
+     * @param rowSize the size of rowkey
      * @return the string
      */
-    String getBatchDeleteLockSql(String lockTable, String paramPlaceHold);
+    String getBatchDeleteLockSql(String lockTable, int rowSize);
 
     /**
      * Get batch delete lock sql string.
@@ -69,10 +69,10 @@ public interface LockStoreSql {
      * Get batch delete lock sql string.
      *
      * @param lockTable      the lock table
-     * @param paramPlaceHold the param place hold
+     * @param branchSize the size of branch
      * @return the string
      */
-    String getBatchDeleteLockSqlByBranchs(String lockTable, String paramPlaceHold);
+    String getBatchDeleteLockSqlByBranchs(String lockTable, int branchSize);
 
     /**
      * Get query lock sql string.
@@ -86,10 +86,10 @@ public interface LockStoreSql {
      * Get check lock sql string.
      *
      * @param lockTable      the lock table
-     * @param paramPlaceHold the param place hold
+     * @param rowSize the size of rowkey
      * @return the string
      */
-    String getCheckLockableSql(String lockTable, String paramPlaceHold);
+    String getCheckLockableSql(String lockTable, int rowSize);
 
     /**
      * get batch update status lock by global sql

--- a/core/src/test/java/io/seata/core/store/db/sql/lock/LockStoreSqlFactoryTest.java
+++ b/core/src/test/java/io/seata/core/store/db/sql/lock/LockStoreSqlFactoryTest.java
@@ -40,6 +40,18 @@ public class LockStoreSqlFactoryTest {
 
     private static String BRANCH_TABLE = "branch_table";
 
+    private static String EXPECT_CHECK_GLOBAL_LOCKABLE_SQL = "select xid, transaction_id, branch_id, resource_id, table_name, pk, row_key, gmt_create, gmt_modified,status from " + GLOBAL_TABLE + " where row_key in ( ?,?,? ) order by status desc ";
+    
+    private static String EXPECT_CHECK_BRANCH_LOCKABLE_SQL = "select xid, transaction_id, branch_id, resource_id, table_name, pk, row_key, gmt_create, gmt_modified,status from " + BRANCH_TABLE + " where row_key in ( ?,?,? ) order by status desc ";
+    
+    private static String EXPECT_BATCH_GLOBAL_DELETE_LOCK_SQL = "delete from " + GLOBAL_TABLE + " where xid = ? and (row_key in ( ?,?,? )) ";
+    
+    private static String EXPECT_BATCH_BRANCH_DELETE_LOCK_SQL = "delete from " + BRANCH_TABLE + " where xid = ? and (row_key in ( ?,?,? )) ";
+  
+    private static String EXPECT_BATCH_GLOBAL_DELETE_LOCK_BY_BRANCHS_SQL = "delete from " + GLOBAL_TABLE + " where xid = ? and (branch_id in ( ?,?,? )) ";
+  
+    private static String EXPECT_BATCH_BRANCH_DELETE_LOCK_BY_BRANCHS_SQL = "delete from " + BRANCH_TABLE + " where xid = ? and (branch_id in ( ?,?,? )) ";
+     
     @Test
     public void mysqlLockTest() {
         String sql;
@@ -56,10 +68,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_SQL,sql);
+        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_SQL,sql);
 
         // Get batch delete lock sql string.
         sql = MYSQL_LOCK_STORE.getBatchDeleteLockSqlByBranch(GLOBAL_TABLE);
@@ -68,10 +80,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_BY_BRANCHS_SQL,sql);
+        sql = MYSQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_BY_BRANCHS_SQL,sql);
 
         // Get query lock sql string.
         sql = MYSQL_LOCK_STORE.getQueryLockSql(GLOBAL_TABLE);
@@ -80,10 +92,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get check lock sql string.
-        sql = MYSQL_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = MYSQL_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = MYSQL_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_GLOBAL_LOCKABLE_SQL,sql);
+        sql = MYSQL_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_BRANCH_LOCKABLE_SQL,sql);
 
     }
 
@@ -103,10 +115,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_SQL,sql);
+        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_SQL,sql);
 
         // Get batch delete lock sql string.
         sql = ORACLE_LOCK_STORE.getBatchDeleteLockSqlByBranch(GLOBAL_TABLE);
@@ -115,10 +127,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_BY_BRANCHS_SQL,sql);
+        sql = ORACLE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_BY_BRANCHS_SQL,sql);
 
         // Get query lock sql string.
         sql = ORACLE_LOCK_STORE.getQueryLockSql(GLOBAL_TABLE);
@@ -127,10 +139,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get check lock sql string.
-        sql = ORACLE_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = ORACLE_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = ORACLE_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_GLOBAL_LOCKABLE_SQL,sql);
+        sql = ORACLE_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_BRANCH_LOCKABLE_SQL,sql);
     }
 
     @Test
@@ -149,10 +161,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_SQL,sql);
+        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_SQL,sql);
 
         // Get batch delete lock sql string.
         sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSqlByBranch(GLOBAL_TABLE);
@@ -161,10 +173,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_BY_BRANCHS_SQL,sql);
+        sql = POSTGRESQL_LOCK_STORE.getBatchDeleteLockSqlByBranchs(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_BY_BRANCHS_SQL,sql);
 
         // Get query lock sql string.
         sql = POSTGRESQL_LOCK_STORE.getQueryLockSql(GLOBAL_TABLE);
@@ -173,10 +185,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get check lock sql string.
-        sql = POSTGRESQL_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = POSTGRESQL_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = POSTGRESQL_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_GLOBAL_LOCKABLE_SQL,sql);
+        sql = POSTGRESQL_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_BRANCH_LOCKABLE_SQL,sql);
     }
 
     @Test
@@ -195,10 +207,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = H2_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = H2_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = H2_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_SQL,sql);
+        sql = H2_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_SQL,sql);
 
         // Get batch delete lock sql string.
         sql = H2_LOCK_STORE.getBatchDeleteLockSqlByBranch(GLOBAL_TABLE);
@@ -207,10 +219,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = H2_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = H2_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = H2_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_BY_BRANCHS_SQL,sql);
+        sql = H2_LOCK_STORE.getBatchDeleteLockSqlByBranchs(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_BY_BRANCHS_SQL,sql);
 
         // Get query lock sql string.
         sql = H2_LOCK_STORE.getQueryLockSql(GLOBAL_TABLE);
@@ -219,10 +231,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get check lock sql string.
-        sql = H2_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = H2_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = H2_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_GLOBAL_LOCKABLE_SQL,sql);
+        sql = H2_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_BRANCH_LOCKABLE_SQL,sql);
     }
 
     @Test
@@ -241,10 +253,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_SQL,sql);
+        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_SQL,sql);
 
         // Get batch delete lock sql string.
         sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSqlByBranch(GLOBAL_TABLE);
@@ -253,10 +265,10 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get batch delete lock sql string.
-        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_GLOBAL_DELETE_LOCK_BY_BRANCHS_SQL,sql);
+        sql = OCEANBASE_LOCK_STORE.getBatchDeleteLockSqlByBranchs(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_BATCH_BRANCH_DELETE_LOCK_BY_BRANCHS_SQL,sql);
 
         // Get query lock sql string.
         sql = OCEANBASE_LOCK_STORE.getQueryLockSql(GLOBAL_TABLE);
@@ -265,9 +277,9 @@ public class LockStoreSqlFactoryTest {
         Assertions.assertNotNull(sql);
 
         // Get check lock sql string.
-        sql = OCEANBASE_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, "1");
-        Assertions.assertNotNull(sql);
-        sql = OCEANBASE_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, "1");
-        Assertions.assertNotNull(sql);
+        sql = OCEANBASE_LOCK_STORE.getCheckLockableSql(GLOBAL_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_GLOBAL_LOCKABLE_SQL,sql);
+        sql = OCEANBASE_LOCK_STORE.getCheckLockableSql(BRANCH_TABLE, 3);
+        Assertions.assertEquals(EXPECT_CHECK_BRANCH_LOCKABLE_SQL,sql);
     }
 }

--- a/server/src/main/java/io/seata/server/storage/db/lock/LockStoreDataBaseDAO.java
+++ b/server/src/main/java/io/seata/server/storage/db/lock/LockStoreDataBaseDAO.java
@@ -23,7 +23,6 @@ import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
-import java.util.StringJoiner;
 import java.util.stream.Collectors;
 import javax.sql.DataSource;
 import io.seata.common.exception.DataAccessException;
@@ -120,13 +119,9 @@ public class LockStoreDataBaseDAO implements LockStore {
                 conn.setAutoCommit(false);
             }
             //check lock
-            StringJoiner sj = new StringJoiner(",");
-            for (int i = 0; i < lockDOs.size(); i++) {
-                sj.add("?");
-            }
             boolean canLock = true;
             //query
-            String checkLockSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getCheckLockableSql(lockTable, sj.toString());
+            String checkLockSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getCheckLockableSql(lockTable, lockDOs.size());
             ps = conn.prepareStatement(checkLockSQL);
             for (int i = 0; i < lockDOs.size(); i++) {
                 ps.setString(i + 1, lockDOs.get(i).getRowKey());
@@ -224,12 +219,8 @@ public class LockStoreDataBaseDAO implements LockStore {
             conn = lockStoreDataSource.getConnection();
             conn.setAutoCommit(true);
 
-            StringJoiner sj = new StringJoiner(",");
-            for (int i = 0; i < lockDOs.size(); i++) {
-                sj.add("?");
-            }
             //batch release lock
-            String batchDeleteSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getBatchDeleteLockSql(lockTable, sj.toString());
+            String batchDeleteSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getBatchDeleteLockSql(lockTable, lockDOs.size());
             ps = conn.prepareStatement(batchDeleteSQL);
             ps.setString(1, lockDOs.get(0).getXid());
             for (int i = 0; i < lockDOs.size(); i++) {
@@ -272,10 +263,8 @@ public class LockStoreDataBaseDAO implements LockStore {
         try {
             conn = lockStoreDataSource.getConnection();
             conn.setAutoCommit(true);
-            StringJoiner sj = new StringJoiner(",");
-            branchIds.forEach(branchId -> sj.add("?"));
             //batch release lock by branch list
-            String batchDeleteSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getBatchDeleteLockSqlByBranchs(lockTable, sj.toString());
+            String batchDeleteSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getBatchDeleteLockSqlByBranchs(lockTable, branchIds.size());
             ps = conn.prepareStatement(batchDeleteSQL);
             ps.setString(1, xid);
             for (int i = 0; i < branchIds.size(); i++) {
@@ -396,13 +385,8 @@ public class LockStoreDataBaseDAO implements LockStore {
         PreparedStatement ps = null;
         ResultSet rs = null;
         try {
-            StringJoiner sj = new StringJoiner(",");
-            for (int i = 0; i < lockDOs.size(); i++) {
-                sj.add("?");
-            }
-
             //query
-            String checkLockSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getCheckLockableSql(lockTable, sj.toString());
+            String checkLockSQL = LockStoreSqlFactory.getLogStoreSql(dbType).getCheckLockableSql(lockTable, lockDOs.size());
             ps = conn.prepareStatement(checkLockSQL);
             for (int i = 0; i < lockDOs.size(); i++) {
                 ps.setString(i + 1, lockDOs.get(i).getRowKey());

